### ISSUE TITLE
"Atomic commit failed (modeset): Invalid argument" after switching to Gnome and backward

### ISSUE THREAD

**Issue by versusvoid** [swaywm/wlroots#204]
Reproduced as (assuming Gnome is already running)
1. Start rootston from tty
2. Switch to Gnome
Screen blinks. Looks like Gnome makes clean DRM initialization.
3. Switch back
Gnome interface will be left frozen on screen.

Rootston log says:
```
2017-10-05 07:48:52 - [backend/drm/backend.c:80] DRM fd paused
2017-10-05 07:48:55 - [backend/drm/backend.c:65] DRM fd resumed
2017-10-05 07:48:55 - [backend/drm/atomic.c:57] VGA-1: Atomic commit failed (modeset): Invalid argument
2017-10-05 07:48:55 - [backend/drm/atomic.c:57] HDMI-A-1: Atomic commit failed (modeset): Invalid argument
```

Not reproduced with _weston_ or _xfwm4_. But when switching from Gnome to _weston_ or _xfwm4_ screen also blinks, so maybe they detect such situations and make clean DRM initialization too.

**Comment by ascent12** [comment]
Is this using logind or the direct session?

**Comment by versusvoid** [comment]
Logind

**Comment by emersion** [comment]
rootston freezes if I have two monitors, but with only one I can switch between rootston and gnome without this issue.

**Comment by versusvoid** [comment]
Yes, and screen does not blink when switching to Gnome with only one monitor.

**Comment by versusvoid** [comment]
Not reproduced on legacy DRM.

**Comment by versusvoid** [comment]
After commenting out
```cpp
atomic_add(&atom, conn->id, conn->props.crtc_id, crtc->id);
```
in `atomic_crtc_pageflip()` commit produce no error and allows switching. Though screen does not blink when switching back to rootston.

**Comment by ascent12** [comment]
That's not a viable solution. That's just being lucky that Gnome happened to pick the CRTC that we did.
You have to go through a full modeset (which causes the temporary blinking) when you get switched to, as you just can't guarantee that the CRTCs and connectors are set up the way you need them to be.

I have a vague idea what might be causing this, but I've haven't really had the time to investigate it properly.

**Comment by versusvoid** [comment]
Enabled DRM debug in kernel and got following:
```
kernel: [drm:intel_atomic_check [i915]] rejecting invalid cloning configuration
kernel: [drm:intel_dump_pipe_config [i915]] [CRTC:32:pipe A][failed]
kernel: [drm:intel_dump_pipe_config [i915]] cpu_transcoder: A, pipe bpp: 18, dithering: 0
kernel: [drm:intel_dump_pipe_config [i915]] audio: 0, infoframes: 0
kernel: [drm:intel_dump_pipe_config [i915]] requested mode:
kernel: [drm:drm_mode_debug_printmodeline [drm]] Modeline 0:"1600x900" 60 110000 1600 1664 1706 2010 900 903 906 912 0x48 0xa
kernel: [drm:intel_dump_pipe_config [i915]] adjusted mode:
kernel: [drm:drm_mode_debug_printmodeline [drm]] Modeline 0:"1600x900" 60 110000 1600 1664 1706 2010 900 903 906 912 0x48 0xa
kernel: [drm:intel_dump_pipe_config [i915]] crtc timings: 110000 1600 1664 1706 2010 900 903 906 912, type: 0x48 flags: 0xa
kernel: [drm:intel_dump_pipe_config [i915]] port clock: 0, pipe src size: 1600x900, pixel rate 0
kernel: [drm:intel_dump_pipe_config [i915]] pch pfit: pos: 0x00000000, size: 0x00000000, disabled
kernel: [drm:intel_dump_pipe_config [i915]] ips: 0, double wide: 0
kernel: [drm:hsw_dump_hw_state [i915]] dpll_hw_state: wrpll: 0x0 spll: 0x94000000
kernel: [drm:intel_dump_pipe_config [i915]] planes on this crtc
kernel: [drm:intel_dump_pipe_config [i915]] [PLANE:26:primary A] FB:85, fb = 1920x1080 format = XR24 little-endian (0x34325258)
kernel: [drm:intel_dump_pipe_config [i915]] [PLANE:28:sprite A] disabled, scaler_id = 0
kernel: [drm:intel_dump_pipe_config [i915]] [PLANE:30:cursor A] FB:82, fb = 256x256 format = AR24 little-endian (0x34325241)
```
Digging further

**Comment by versusvoid** [comment]
Adding this
```patch
diff --git a/backend/drm/backend.c b/backend/drm/backend.c
index d7e2e61..8b7e67e 100644
--- a/backend/drm/backend.c
+++ b/backend/drm/backend.c
@@ -13,6 +13,7 @@
 #include <wlr/util/log.h>
 #include <wlr/egl.h>
 #include "backend/drm/drm.h"
+#include "backend/drm/util.h"
 
 static bool wlr_drm_backend_start(struct wlr_backend *backend) {
 	struct wlr_drm_backend *drm = (struct wlr_drm_backend *)backend;
@@ -64,6 +65,50 @@ static void session_signal(struct wl_listener *listener, void *data) {
 	if (session->active) {
 		wlr_log(L_INFO, "DRM fd resumed");
 
+		drmModeAtomicReqPtr req = drmModeAtomicAlloc();
+		for (size_t i = 0; i < drm->outputs->length; ++i) {
+			struct wlr_drm_connector *conn = drm->outputs->items[i];
+			if (conn->state != WLR_DRM_CONN_CONNECTED) {
+				continue;
+			}
+
+			struct wlr_drm_crtc *crtc = conn->crtc;
+			struct wlr_drm_plane *plane = crtc->primary;
+
+			struct gbm_bo *bo = wlr_drm_surface_get_front(
+				drm->parent ? &plane->mgpu_surf : &plane->surf);
+			uint32_t fb_id = get_fb_for_bo(bo);
+
+			struct wlr_drm_mode *mode = (struct wlr_drm_mode *)conn->output.current_mode;
+			if (crtc->mode_id) {
+				drmModeDestroyPropertyBlob(drm->fd, crtc->mode_id);
+			}
+			drmModeModeInfo *drm_mode = &mode->drm_mode;
+
+			if (drmModeCreatePropertyBlob(drm->fd, drm_mode, sizeof(*drm_mode), &crtc->mode_id)) {
+				wlr_log_errno(L_ERROR, "Unable to create property blob");
+				assert(false);
+			}
+
+			drmModeAtomicAddProperty(req, conn->id, conn->props.crtc_id, conn->crtc->id);
+			drmModeAtomicAddProperty(req, crtc->id, crtc->props.mode_id, crtc->mode_id);
+			drmModeAtomicAddProperty(req, crtc->id, crtc->props.active, 1);
+
+			drmModeAtomicAddProperty(req, plane->id, plane->props.src_x, 0);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.src_y, 0);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.src_w, plane->surf.width << 16);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.src_h, plane->surf.height << 16);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.crtc_w, plane->surf.width);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.crtc_h, plane->surf.height);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.fb_id, fb_id);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.crtc_id, crtc->id);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.crtc_x, 0);
+			drmModeAtomicAddProperty(req, plane->id, plane->props.crtc_y, 0);
+		}
+		drmModeAtomicCommit(drm->fd, req, DRM_MODE_ATOMIC_ALLOW_MODESET, NULL);
+		drmModeAtomicFree(req);
+
+
 		for (size_t i = 0; i < drm->outputs->length; ++i) {
 			struct wlr_drm_connector *conn = drm->outputs->items[i];
 			wlr_drm_connector_start_renderer(conn);
```

Also solves the problem, so my guess is that current atomic isn't atomic enough, since wlroots sends one modeset request per output.

**Comment by ascent12** [comment]
Again, not a valid solution, as you're using atomic operations when they're not guaranteed to exist.
Due to it complaining about `invalid cloning configuration`, my guess would be to try something like
```c
for (size_t i = 0; i < drm->outputs->length; ++i) {
    struct wlr_drm_connector *conn = drm->outputs->items[i];
    if (conn->state != WLR_DRM_CONN_CONNECTED) {
        continue;
    }

    drmModeSetCrtc(drm->fd, 0, 0, 0, 0, &conn->id, 1, NULL);
}```

**Comment by versusvoid** [comment]
Never meant to be, just an illustration.
So yeah, problem is with Gnome switching connector-crtc binding. I tried manually swapping `wlr_drm_crtc` in `wlr_drm_connector`s and modeset was successful.

There are two solutions I see:
1. Make modeset atomic across all outputs (like in the patch above, but hidden behind interface)
2. Reread DRM state (maybe modify `wlr_drm_scan_connectors()`) and sync our DRM's state representation (inside `wlr_drm_connector` and below) with it.

I tested first approach. Second may have some obstacles I'm not seeing right now.

---

Adding your code after
```cpp
wlr_log(L_INFO, "DRM fd resumed");
```
haven't resolve the problem for me.

---

And it's clear now, why removing
```cpp
atomic_add(&atom, conn->id, conn->props.crtc_id, crtc->id);
```
seemingly helped. rootston didn't tired switching CRTC for connectors and set same resolutions for wrong planes.
